### PR TITLE
Pipeline Operator proposal

### PR DESCRIPTION
**Pull request by hzoo**
Ref https://github.com/babel/proposals/issues/29
Taken from https://github.com/babel/babel/pull/3159 by @gilbert
Spec: https://github.com/tc39/proposal-pipeline-operator/pull/51/files

### need more tests
- [x] precedence
- [x] with arrow function `10 |> x => x + 1;`
- [x] chaining `a |> b |> c`
- [x] multline

```js
let result = "hello"
  |> doubleSay
  |> capitalize
  |> exclaim;
```

```js
let newScore = person.score
  |> double
  |> _ => add(7, _)
  |> _ => boundScore(0, 100, _);
```

### CODE REVIEW

File: src/tokenizer/index.js [hunk 1]
```diff
@@ -411,9 +411,17 @@ export default class Tokenizer extends LocationParser {
     const next = this.input.charCodeAt(this.state.pos + 1);
     if (next === code)
       return this.finishOp(code === 124 ? tt.logicalOR : tt.logicalAND, 2);
+    if (code === 124) {
+      // '|>'
+      if (next === 62) {
+        this.expectPlugin("pipelineOperator");
```
Comment (jridgewell) [inline src/tokenizer/index.js]: Don't we usually expect in the not-tokenizer part?

File: src/tokenizer/index.js [hunk 2]
```diff
@@ -411,9 +411,17 @@ export default class Tokenizer extends LocationParser {
     const next = this.input.charCodeAt(this.state.pos + 1);
```
Comment (jridgewell) [inline src/tokenizer/index.js]: Please rename the function.

File: test/fixtures/experimental/pipeline-operator/base/actual.js [hunk 3]
```diff
@@ -0,0 +1 @@
+a |> b
```
Comment (jridgewell) [inline test/fixtures/experimental/pipeline-operator/base/actual.js]: More examples, especially arrows.

File: src/tokenizer/index.js (hunk 2 above)
Comment (hzoo) [reply src/tokenizer/index.js]: this is ok now

File: test/fixtures/experimental/pipeline-operator/base/actual.js (hunk 3 above)
Comment (hzoo) [reply test/fixtures/experimental/pipeline-operator/base/actual.js]: updated

File: src/tokenizer/types.js [hunk 4]
```diff
@@ -61,7 +61,7 @@ export class TokenType {
     this.isAssign = !!conf.isAssign;
     this.prefix = !!conf.prefix;
     this.postfix = !!conf.postfix;
-    this.binop = conf.binop || null;
+    this.binop = conf.binop === 0 ? 0 : conf.binop || null;
```
Comment (littledan) [inline src/tokenizer/types.js]: This will be nice with ?? :)

File: src/parser/expression.js [hunk 5]
```diff
@@ -290,6 +290,13 @@ export default class ExpressionParser extends LValParser {
 
         const startPos = this.state.start;
         const startLoc = this.state.startLoc;
+
+        if (node.operator === "|>") {
+          this.expectPlugin("pipelineOperator");
+          // Support syntax such as 10 |> x => x + 1
+          this.state.potentialArrowAt = startPos;
```
Comment (littledan) [inline src/parser/expression.js]: I don't think the current draft spec has this logic. I couldn't figure out how to put it into the grammar. Right now, `x |> y => z` would be a syntax error. But it's clearly useful--i'll think more about the spec.

Comment (hzoo) [reply src/parser/expression.js]: Good to know

File: src/tokenizer/types.js (hunk 4 above)
Comment (hzoo) [reply src/tokenizer/types.js]: Yes, should be easy to add/use here (would be good to find this kind of code on github/codebase and then do a codemod)

**Pull request merged**